### ISSUE TITLE
Backpressure

### ISSUE THREAD

**Issue by mitar** [apollostack/apollo#16]
One thing I really miss in DDP is backpressure support. So that the way client could say back to the server: please slow down, I cannot process changes as quickly as you are sending them. Or even, please batch them together. I do not need every change, you can aggregate changes over the same field together and send them every X seconds.

I think this is one thing node.js learned a lot with their multiple iterations of streams. And how they were adding backpressure to them.

A good talk about one similar project with backpressure is this one: http://www.infoq.com/presentations/streaming-data-back-pressure

**Comment by arunoda** [comment]
I think this is enabled by default with apollo. Basically, this is what happens.

* Invalidation server sends the documents which are invalidated (basically, they are just a set of ids)
* Then client will directly re-fetch data related from the Data server. (GraphQL endpoint)

So, client can decide how often/fast, it needs to fetch data.

**Comment by stubailo** [comment]
Yes, I think it is important that we don't have an architecture where the server always pushes stuff without letting the client reject it.

**Comment by MasterJames** [comment]
It forces fields if I'm not mistaken.

**Comment by stubailo** [comment]
> It forces fields if I'm not mistaken.

What do you mean?

**Comment by sebakerckhof** [comment]
I think that since the initial design document focuses so much on a stateless, easy to test server, we shouldn't implement backpressure into the server even for push streams. If we need this, it could be done by means of a buffer (possibly with aggregation) on the client IMO. The use case where even the client buffer can't keep up, is probably too rare to care about.

**Comment by mitar** [comment]
Sorry, but this is bad thinking. If Meteor is really trying to move away from small projects into enterprise, handling edge cases under high load is important.

**Comment by arunoda** [comment]
> Sorry, but this is bad thinking.

What is bad thinking?
I think, the point here is that Apollo brings manageable scalability for any kind of app.

**Comment by mitar** [comment]
> What is bad thinking?

This:

> If we need this, it could be done by means of a buffer (possibly with aggregation) on the client IMO.

**Comment by arunoda** [comment]
Okay.

**Comment by sebakerckhof** [comment]
> Sorry, but this is bad thinking. If Meteor is really trying to move away from small projects into enterprise, handling edge cases under high load is important.

1) Enterprise projects do not necessarily have higher load than smaller projects, in fact, there is no causality at all here.

2) The buffer on the client side could be part of the library. This allows one to have backpressure from the view layer to the client side cache, where the bottleneck usually is. You could have backpressure from the client side cache to the server, but that complicates the server quite a bit and the initial design doc of apollo is around making the server state-free, easy to test etc. 
The edge case I'm talking about is if the rate of updates coming into the client side cache is too high. This would be very rare and I actually do not think meteor/apollo should handle every edge case. It should try to do good for 99% of the use cases.

**Comment by MasterJames** [comment]
as to 'forced fields' I meant more about reducing the load in other ways (discarding unneeded data).
Got the thought from your (somewhat related) comments actually.
https://www.youtube.com/watch?v=VQOsDlEnN4U&t=3152

The GraphQL logic does this reduction in many ways too (also from video), like avoiding multiple collection calls to boil down to a few records and really you only want the most popular websiteUrl field.
At that point you're pretty darn efficient and then it's time to spin up more hardware.

The server and client should both adjust automatically to aggregation mode when either/or detects back-pressure.

_Rant: (With something like more streamed data, there are expectations animation splines even)_
It seems from my perspective the same active rules of interactive 3D come to play in reactive data. This example use case is that of a game engine. Depending of FPS (frames per second) you can turn off shadows, change alias mode, lower texture map resolution, all to satisfy a rule of preference for speed [higher fps], over image quality depending on your hardware (or even zoom level).
This can be automated to find optimal settings for a given client. If their frame rate is 15 a second then you don't need data update rates of 60 like maybe someone else can handle.
Then you can choose to talk about vectors or value targets in the future (aka animation curves). Rate of Change projections or prediction of data changes to come can be assumed or are in fact known or knowable (in the case of movement/momentum etc.). This might happen when the FPS is fine but the data updates are slow. You see the key-frame comes with a time-stamp so if it's too old either side can drop it but the missed change will be smoothed out by estimations or animation curves (momentum or splines essentially).

Anyway the point is interactive systems need real-time performance monitoring and feature-tweaking per client in an automated way that can be guided by preferences of the user/client and their hardware's capabilities.
It's the developer's need to manage those things, but I think some measure of this built into the underlying system is also needed.
Look at another example converting json to bson takes clock cycles if the connection is super fast maybe you don't want to waste clock cycles doing compression on both sides, because your cycles are tight/limited.
Automated Real-time performance tweaking with a quick low priority test suite could reveal otherwise humanly unknowable improvements (ISD intelligent software design). So in the spectrum of use cases on bandwidth versas core cycles compression would be one factor affecting that correlation to be tested and monitored per client some in an initial step, and some regularly as a low priority when possible. Some would need it and others wouldn't, and not just the use case of an app vs another one but from one client to another in a given app. Like 3D on Mobile vs highend Desktop, same app totally different settings needed for performance optimization, that can be adjusted based on real-time needs. Even models can adjust their detail/performance-hit based on distance from camera. 
[Anyway I've ranted enough. You can take from it what you will. Automated Optimization is a developer problem, but you can find ways to integrate these ideas on the lower system levels some I'm sure.]

**Comment by mitar** [comment]
It seems streams are becoming a web standard. Probably Apollo should then use them?
* https://jakearchibald.com/2016/streams-ftw/
* https://streams.spec.whatwg.org/

**Comment by stubailo** [comment]
> It seems streams are becoming a web standard.

I've never heard of them outside of the links you posted, is there evidence that they are gaining traction among developers and browsers that aren't Chrome?

**Comment by mitar** [comment]
> Response objects, as defined by the fetch spec, let you read the response as a variety of formats, but response.body gives you access to the underlying stream. response.body is supported in the current stable version of Chrome.

**Comment by stubailo** [comment]
> is there evidence that they are gaining traction among developers and browsers that aren't Chrome?

**Comment by helfer** [comment]
I think if we have a design where clients can essentially specify how frequently they want updates, that could be used to implement backpressure, or am I mistaken? It's unlikely that it will be in the early versions of Apollo, but eventually it will be part of Apollo.

Regarding streams: Shouldn't that be a discussion in a separate issue?

**Comment by mitar** [comment]
> is there evidence that they are gaining traction among developers and browsers that aren't Chrome?

Oh, this is very still in progress. I just mentioned it for a record here.
